File: 002/evolutionary.py
from random import randrange

class evolution():

    def __init__(self):
        self.target = list('METHINKS IT IS LIKE A WEASEL')
        self.charset = list('ABCDEFGHIJKLMNOPQRSTUVWXYZ ')

        self.parent = []
        self.parentMatches = []
        self.matches = 0

    def randomParent(self):
        # Sorteia a primeira 'especie'
        for i in range(len(self.target)):
            randomValue = randrange(len(self.charset))
            self.parent.append(self.charset[randomValue])
            self.parentMatches.append(False)
        
    def fitness(self, mutant):
        # Ve o quao perto esta o mutant do target, se for mais perto que
        # o parent, troca de lugar com o parent.
        # Retornos:
        #   True  -- identico ao esperado;
        #   False -- mais proximo da resp;
        #   None  -- sem avanco.
        auxMatches = []
        cont = 0
        for i in range(len(self.target)):
            if mutant[i] == self.target[i]:
                auxMatches.append(True)
                cont += 1
            else:
                auxMatches.append(False)

        if (cont == len(self.target)):
            return True

        if (cont >= self.matches):
            self.parent = mutant
            self.parentMatches = auxMatches
            self.matches = cont
            return False
        
        return None

    def mutate(self):
        # Sorteia mudancas no parent e retorna o mutante
        mutate = []
        for i in range(len(self.parent)):
            if not self.parentMatches[i]:
                randomValue = randrange(len(self.charset))
                mutate.append(self.charset[randomValue])
            else:
                mutate.append(self.parent[i])
        
        return mutate

    def builder(self):
        print('Texto alvo: {0}'.format(self.__passListToString(self.target)))

        # Define a primeira especie
        self.randomParent()
        self.fitness(self.parent)
        print('Primeira esp: {0}'.format(self.__passListToString(self.parent)))
    
        cont = 0
        while True:
            cont += 1
            mutant = self.mutate()
            isEquals = self.fitness(mutant)
            if isEquals == None:
                #mutante sem resultados
                pass
            elif isEquals:
                #igual
                print('tentativa {0}: SUCESSO'.format(cont))
                break
            elif not isEquals:
                #melhor que o pai
                percent = self.__calcPercent()
                print('tentativa {0}: {1:.2f}%'.format(cont, percent))

    def __passListToString(self, list):
        return ''.join(map(str, list))
    
    def __calcPercent(self):
        return (self.matches / 28) * 100

e= evolution()
e.builder()

File: 002/evolutionary.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#define TEXT_SIZE 28
#define LETTERS_SIZE 27

int getRandom();

void passString1ToString2();

float calcPercent();

/*
  Sorteia a primeira 'especie'.
*/
int randomParent();

/*
  Ve o quao perto esta o mutant do target, se for mais perto que
  o parent, troca de lugar com o parent.
  Retornos:
    1  -- identico ao esperado;
    2 -- mais proximo da resp;
    0  -- sem avanco.
*/
int fitness();

/*
  Sorteia mudancas no parent e retorna o mutante.
*/
void getMutate();

int main()
{
  srand(time(NULL));

  char target[TEXT_SIZE + 1] = "METHINKS IT IS LIKE A WEASEL";
  char charset[LETTERS_SIZE + 1] = "ABCDEFGHIJKLMNOPQRSTUVWXYZ ";

  char parent[TEXT_SIZE + 1];
  char parentMatches[TEXT_SIZE + 1];
  int matches = 0;

  char mutate[TEXT_SIZE + 1];

  randomParent(parent, charset, parentMatches);
  int isEquals = fitness(parent, &matches, target, parent, parentMatches);

  printf("Texto alvo: %s\n", target);
  printf("Primeiro esp: %s\n", parent);
  int cont = 0;
  while (1)
  {
    cont += 1;
    getMutate(mutate, parent, parentMatches, charset);
    isEquals = fitness(mutate, &matches, target, parent, parentMatches);
    if (isEquals == 1)
    {
      //igual
      printf("tentativa %d: SUCESSO\n", cont);
      break;
    }
    else if (isEquals == 2)
    {
      //melhor que o pai
      printf("tentativa %d: %.2f%% \n", cont, calcPercent(matches, TEXT_SIZE));
    }
  }

  return 1;
}

int getRandom(int initValue, int endValue)
{
  return (rand() % endValue + initValue);
}

void passString1ToString2(char *s1, char *s2)
{
  for (int i = 0; i < TEXT_SIZE; i++)
  {
    s2[i] = s1[i];
  }
}

float calcPercent(int val, int max)
{
  return ((float)val / (float)max) * 100.0;
}

int randomParent(char *parent, char *charset, char *parentMatches)
{
  for (int i = 0; i < TEXT_SIZE; i++)
  {
    int randomValue = getRandom(0, LETTERS_SIZE);
    parent[i] = charset[randomValue];
    parentMatches[i] = 'f';
  }
  parent[TEXT_SIZE] = '\0';
  parentMatches[TEXT_SIZE] = '\0';
}

int fitness(char *mutant, int *matches, char *target, char *parent,
            char *parentMatches)
{
  char auxMatches[TEXT_SIZE + 1];
  int cont = 0;
  for (int i = 0; i < TEXT_SIZE; i++)
  {
    if (mutant[i] == target[i])
    {
      auxMatches[i] = 't';
      cont++;
    }
    else
    {
      auxMatches[i] = 'f';
    }
  }
  auxMatches[TEXT_SIZE] = '\0';

  if (cont == TEXT_SIZE)
  {
    passString1ToString2(auxMatches, parentMatches);
    return 1;
  }

  if (cont >= *matches)
  {
    passString1ToString2(mutant, parent);
    passString1ToString2(auxMatches, parentMatches);
    *matches = cont;
    return 2;
  }

  return 0;
}

void getMutate(char *mutate, char *parent, char *parentMatches, char *charset)
{
  // Sorteia mudancas no parent e retorna o mutante
  for (int i = 0; i < TEXT_SIZE; i++)
  {
    if (parentMatches[i] == 'f')
    {
      int randomValue = getRandom(0, LETTERS_SIZE);
      mutate[i] = charset[randomValue];
    }
    else
    {
      mutate[i] = parent[i];
    }
  }
  mutate[TEXT_SIZE] = '\0';
}